core/embed/rust/src/ui/model_tt/bootloader/mod.rs

```diff
@@ -511,3 +531,158 @@ extern "C" fn screen_welcome() {
+        }
+    } else {
+        None
+    };
+
+    let vendor_str = if vendor_str.is_null() {
+        None
+    } else {
+        unsafe { from_c_array(vendor_str, vendor_str_len as _) }
+    };
+
+    let version = if version.is_null() {
+        None
+    } else {
+        unsafe { from_c_str(version) }
+    };
+
+    fadeout();
+
+    const IMG_SIZE: i16 = 120;
+
+    let img_pos = if vendor_str.is_some() {
+        Point::new((screen().width() - IMG_SIZE) / 2, 30)
+    } else {
+        Point::new(
+            (screen().width() - IMG_SIZE) / 2,
+            (screen().height() - IMG_SIZE) / 2,
+        )
+    };
+
+    let bg_color = if (vtrust & (VTRUST_RED as u16)) == 0 {
+        RED
+    } else {
+        BLACK
+    };
+
+    display::rect_fill(screen(), bg_color);
+
+    if let Some(img) = img {
+        let toif = Toif::new(img).unwrap();
+        if toif.format() == ToifFormat::FullColorLE
+            && toif.width() == IMG_SIZE
+            && toif.height() == IMG_SIZE
+        {
+            let img = Image::new(img);
+            img.draw(img_pos, TOP_LEFT);
+        }
+    }
+
+    if let Some(vendor_str) = vendor_str {
+        display::text_center(
+            Point::new(screen().width() / 2, screen().height() - 5 - 50),
+            vendor_str,
+            Font::NORMAL,
+            WHITE,
+            bg_color,
+        );
+        if let Some(ver_str) = version {
+            display::text_center(
+                Point::new(screen().width() / 2, screen().height() - 5 - 25),
+                ver_str,
+                Font::NORMAL,
+                WHITE,
+                bg_color,
+            );
+        }
+    }
+
+    fadein();
+
+    let mut wait = (vtrust as u32 & VTRUST_WAIT) ^ VTRUST_WAIT;
+    match wait {
+        0 => {}
+        1 => {
+            time::sleep(Duration::from_millis(1000));
+        }
+        _ => {
+            while wait > 0 {
+                let mut buff = [0u8; 10];
+                let mut wait_str: String<32> = String::new();
+                wait_str.push_str("starting in ").unwrap();
+                wait_str
+                    .push_str(u32_to_str(wait as _, &mut buff).unwrap())
+                    .unwrap();
+                wait_str.push_str(" s").unwrap();
+
+                display::rect_fill(
+                    Rect::new(
+                        Point::new(0, screen().height() - 5 - 20),
+                        Point::new(screen().width(), screen().height()),
+                    ),
+                    bg_color,
+                );
+                display::text_center(
+                    Point::new(screen().width() / 2, screen().height() - 5),
+                    wait_str.as_str(),
+                    Font::NORMAL,
+                    WHITE,
+                    bg_color,
+                );
+                time::sleep(Duration::from_millis(1000));
+                wait -= 1;
+            }
+        }
+    }
+
+    if (vtrust & (VTRUST_CLICK as u16)) == 0 {
```

Comment:
What is this `vtrust` thing?

Reply:
its from vendor header - contains flags about how should the firmware be trusted. e.g. firmware signed by satoshi labs is trusted and doesn't require user confirmation when starting up, whereas development firmware does.